File: appthemehelper/src/main/java/code/name/monkey/appthemehelper/common/prefs/supportv7/dialogs/ATEPreferenceDialogFragment.java
package code.name.monkey.appthemehelper.common.prefs.supportv7.dialogs;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.util.Log;
import android.view.Window;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.preference.DialogPreference;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;

/**
 * @author Karim Abou Zeid (kabouzeid)
 */
public class ATEPreferenceDialogFragment extends DialogFragment implements DialogInterface.OnClickListener {
    protected static final String ARG_KEY = "key";
    private static final String TAG = "ATEPreferenceDialog";
    private int mWhichButtonClicked;
    private DialogPreference mPreference;

    public static ATEPreferenceDialogFragment newInstance(String key) {
        ATEPreferenceDialogFragment fragment = new ATEPreferenceDialogFragment();
        Bundle b = new Bundle(1);
        b.putString(ARG_KEY, key);
        fragment.setArguments(b);
        return fragment;
    }

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Fragment rawFragment = this.getTargetFragment();
        if (!(rawFragment instanceof DialogPreference.TargetFragment)) {
            throw new IllegalStateException("Target fragment must implement TargetFragment interface");
        } else {
            DialogPreference.TargetFragment fragment = (DialogPreference.TargetFragment) rawFragment;
            String key = this.getArguments().getString(ARG_KEY);
            this.mPreference = fragment.findPreference(key);
        }
    }

    @NonNull
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        FragmentActivity context = this.getActivity();
        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(context,
                com.google.android.material.R.style.ThemeOverlay_MaterialComponents_Dialog_Alert)
                .setTitle(this.mPreference.getDialogTitle())
                .setIcon(this.mPreference.getDialogIcon())
                .setMessage(this.mPreference.getDialogMessage())
                .setPositiveButton(this.mPreference.getPositiveButtonText(), this)
                .setNegativeButton(this.mPreference.getNegativeButtonText(), this);

        this.onPrepareDialogBuilder(builder);
        AlertDialog dialog = builder.create();
        if (this.needInputMethod()) {
            this.requestInputMethod(dialog);
        }
        return dialog;
    }

    public DialogPreference getPreference() {
        return this.mPreference;
    }

    protected void onPrepareDialogBuilder(MaterialAlertDialogBuilder builder) {
    }

    protected boolean needInputMethod() {
        return false;
    }

    private void requestInputMethod(Dialog dialog) {
        Window window = dialog.getWindow();
        window.setSoftInputMode(5);
    }

    @Override
    public void onDismiss(@NonNull DialogInterface dialog) {
        super.onDismiss(dialog);
        Log.i(TAG, "onDismiss: " + mWhichButtonClicked);
        onDialogClosed(mWhichButtonClicked == DialogInterface.BUTTON_POSITIVE);
    }

    public void onDialogClosed(boolean positiveResult) {

    }

    @Override
    public void onClick(DialogInterface dialog, int which) {
        Log.i(TAG, "onClick: " + which);
        mWhichButtonClicked = which;
        onDialogClosed(which == DialogInterface.BUTTON_POSITIVE);
    }
}

File: appthemehelper/src/main/java/code/name/monkey/appthemehelper/common/prefs/supportv7/dialogs/ATEPreferenceDialogFragment.kt
package code.name.monkey.appthemehelper.common.prefs.supportv7.dialogs

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import androidx.fragment.app.DialogFragment
import androidx.preference.DialogPreference
import com.afollestad.materialdialogs.DialogAction
import com.afollestad.materialdialogs.MaterialDialog

/**
 * @author Karim Abou Zeid (kabouzeid)
 */
open class ATEPreferenceDialogFragment : DialogFragment(), MaterialDialog.SingleButtonCallback {
    private var mWhichButtonClicked: DialogAction? = null
    var preference: DialogPreference? = null
        private set

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val rawFragment = this.targetFragment
        if (rawFragment !is DialogPreference.TargetFragment) {
            throw IllegalStateException("Target fragment must implement TargetFragment interface")
        } else {
            val fragment = rawFragment as DialogPreference.TargetFragment?
            val key = this.arguments!!.getString(ARG_KEY)
            this.preference = fragment!!.findPreference(key) as DialogPreference
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val context = this.activity
        val builder = MaterialDialog.Builder(context!!)
                .title(this.preference!!.dialogTitle)
                .icon(this.preference!!.dialogIcon)
                .onAny(this)
                .positiveText(this.preference!!.positiveButtonText)
                .negativeText(this.preference!!.negativeButtonText)

        builder.content(this.preference!!.dialogMessage)
        this.onPrepareDialogBuilder(builder)
        val dialog = builder.build()
        if (this.needInputMethod()) {
            this.requestInputMethod(dialog)
        }

        return dialog
    }

    protected open fun onPrepareDialogBuilder(builder: MaterialDialog.Builder) {}

    protected open fun needInputMethod(): Boolean {
        return false
    }

    private fun requestInputMethod(dialog: Dialog) {
        val window = dialog.window
        window!!.setSoftInputMode(5)
    }

    override fun onClick(dialog: MaterialDialog, which: DialogAction) {
        mWhichButtonClicked = which
    }

    override fun onDismiss(dialog: DialogInterface?) {
        super.onDismiss(dialog)
        onDialogClosed(mWhichButtonClicked == DialogAction.POSITIVE)
    }

    open fun onDialogClosed(positiveResult: Boolean) {

    }

    companion object {
        const val ARG_KEY = "key"

        fun newInstance(key: String): ATEPreferenceDialogFragment {
            val fragment = ATEPreferenceDialogFragment()
            val b = Bundle(1)
            b.putString(ARG_KEY, key)
            fragment.arguments = b
            return fragment
        }
    }
}
